// @vuepress/core/lib/app/resolveAppPages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveAppPages = void 0;
const utils_1 = require("@vuepress/utils");
const page_1 = require("../page");
const log = (0, utils_1.debug)('vuepress:core/app');
/**
 * Resolve pages for vuepress app
 */
const resolveAppPages = async (app) => {
    log('resolveAppPages start');
    // resolve page absolute file paths according to the page patterns
    const pageFilePaths = await (0, utils_1.globby)(app.options.pagePatterns, {
        absolute: true,
        cwd: app.dir.source(),
    });
    // create pages from files
    const pages = await Promise.all(pageFilePaths.map(async (filePath) => {
        const pageOptions = { filePath };
        // plugin hook: extendsPageOptions
        await app.pluginApi.hooks.extendsPageOptions.process(pageOptions, app);
        const page = await (0, page_1.createPage)(app, pageOptions);
        // plugin hook: extendsPage
        await app.pluginApi.hooks.extendsPage.process(page, app);
        return page;
    }));
    // if there is no 404 page, add one
    if (!pages.some((page) => page.path === '/404.html')) {
        pages.push(await (0, page_1.createPage)(app, {
            path: '/404.html',
            frontmatter: {
                layout: '404',
            },
        }));
    }
    log('resolveAppPages finish');
    return pages;
};
exports.resolveAppPages = resolveAppPages;
